### ISSUE TITLE
Try to prevent confusables

### ISSUE THREAD

**Issue by DanielOaks** [oragono/oragono#178]
Unicode distributes a [confusables.txt](https://www.unicode.org/Public/security/latest/confusables.txt) document. We can use that to help us reject confusing names, as well as possibly rejecting names that contain characters from more than one alphabet. Maybe detect when something contains a character confusable with a latin alphabet character, and then disallow latin alphabet chars and those confusing chars from appearing together in names?

Flesh out this system more and work out exactly how we want to do it. Probably make it a subpackage with its' own tests and all.

**Comment by slingamn** [comment]
The current casemapping implementation seems to provide some degree of protection already:

````
nick ｅｖａｎ
:irc.darwin.network NOTICE ｅｖａｎ :Nickname is reserved; you must change it or authenticate to NickServ within 30s
````

**Comment by slingamn** [comment]
`еvan` tricked it, though (that's a U+0435 for the e, ASCII for the others).

This is the most sophisticated approach I've seen: https://github.com/vhf/confusable_homoglyphs

**Comment by slingamn** [comment]
Long term I reckon we should:

1. Port the `confusable_homoglyphs` library to Go and disallow confusable nicknames and channel names. (It may be possible to collaborate with upstream to simplify the porting process across multiple languages.)
1. Allow emoji (and possibly some other glyphs) via the approach in `devel+emoji` (effectively, split the name at the emojis, casefold each piece independently, put everything back together).

**Comment by slingamn** [comment]
This should have been obvious, but there are homoglyphs that are not mixed-script, e.g., "POTATO" and "РОТАТО" (pure Cyrillic), which are not detected by the Python library. (These are the so-called "whole-script confusables.")

Alternate approach: don't try to ban strings that are "inherently" confusable (such as the two variants of potato). Instead, only try to protected reserved nicknames, using [TR 39's skeleton algorithm](http://unicode.org/reports/tr39/#Confusable_Detection): for each reserved nickname, compute its skeleton, and ban new nicknames and account names that have the same skeleton as a reserved nick.

**Comment by slingamn** [comment]
We could keep an index of skeleton-to-nick in the unreserved case as well, and block changes to a new nick if the skeleton collides.

I'm still trying to work out how skeleton interacts with PRECIS casefolding. I think *probably* we can casefold first and then take the skeleton? But it's possible that there are two distinct strings that are confusable, but which are no longer confusable after casefolding.

Since accounts are the primary mechanism for blocking impersonation, I'm thinking about an option to restrict account names to particular character sets.

**Comment by slingamn** [comment]
If we casefold first, we lose some of the necessary information about which strings are visually confusable. So I'm thinking first take the skeleton, then casefold the skeleton (overriding the precis bidi rule, because we expect skeletons to mix scripts).

This has some weird consequences, however: since the skeleton is no longer a function of the casefolded identifier, you could theoretically own an identifier but not own all of its possible casefoldings, because one of them is owned by someone else. I can't actually think of a case where this would happen, though.

Another weird thing is that the official confusables table considers `m` confusable with `rn`, and `0` confusable with `O`. This has the potential to cause conflicts in already-registered account names and usernames. I'm thinking of excepting these via the following pseudocode:

````python
def oragonoSkeleton(name):
    if is_alphanumeric_ascii(name):
        return id
    else:
        return realSkeleton(name)
````

**Comment by slingamn** [comment]
This was implemented in #334, and a bug was fixed in #340. Closing this.
